javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+           case Left(i) => Map(i -> this)
+         }) ++ (regex match {
+             case Left(r) => Map[Int, Node]()
+             case Right(cc) => (Map[Int, Node]() /: cc) {
+               case (mm, c) => mm ++ c.obtainMapping
+             }})
+      }
+    }
+    
+    // (A(B))(C) converted yields
+    // (((A)(B))(C)) == Node(Right(("","")), 0, Right(Seq(
+    //                    Node(Left(1), 1, Right(Seq(
+    //                      Node(Right(("","")), 1, Left("A")),
+    //                      Node(Left(2), 1, Left("B")))),
+    //                    Node(Left(3), 4, Left("C")))))
+    
+    // parseRegex("a") == Node(Right(("","")),Left("a"))
+    // parseRegex("(a)") == Node(Left(1),Left("a"))
+    // parseRegex("(A(B))(C)") = Node(Right(("","")),Right(List(Node(Left(1),Right(List(Node(Right(("","")),Left("A")), Node(Left(2),Left("B"))))), Node(Left(3),Left("C")))))
+    // parseRegex("(?:A)") = Node(Right(("","")),Left("A"))
+    // parseRegex("(?:(A))") = Node(Left(1),Left("a"))
+    // parseRegex("((?:A))") = Node(Left(1),Left("a"))
+    // parseRegex("((a))") = Node(Left(1),Right(List(Node(Left(2),Left(a)))))
+    // parseRegex("(hum)?") = Node(Right((,?)),Right(List(Node(Left(1),Left(hum)))))
+    // parseRegex("(?=hum)") = Node(Right(((?=,))),Left(hum))
+    // parseRegex("(?=h(um))") = Node(Right(((?=,))),Right(List(Node(Right((,)),Left(h)), Node(Left(1),Left(um)))))
+    // parseRegex("(ab){1,2}") = Node(Right((,{1,2})),Right(List(Node(Left(1),Left(ab)))))
+    // parseRegex("(?:ab)*") = Node(Right((,)),Right(List(Node(Right(,*),Left(ab)))))
+    
+    def parseRegex(pattern: String): Node = {
+       // Take care of escaped parentheses \( and \)
+       // Takes care of non-group parenthesis (?:)
+       // Returns a sequence of nodes, the remaining non-parsed, and the next group index
+       def parseClosingParenthesis(s: String) = s.substring(1)
+       def parseUntilEndBrace(s: String): (String, String) = if(s.length == 0) ("", "") else if(s(0) =='}') ("}", s.substring(1)) else {
+         val (inside, after) = parseUntilEndBrace(s.substring(1))
+         (s(0) + inside, after)
+       }
+       def parseUntilEndSquareBracket(s: String): (String, String) = if(s.length == 0) ("", "") else if(s(0) == '\\' && s.length > 1) {
+         val (inside, after) = parseUntilEndSquareBracket(s.substring(2))
+         (s(0) + "" + s(1) + inside, after)
+       } else if(s(0) ==']') ("]", s.substring(1)) else {
+         val (inside, after) = parseUntilEndSquareBracket(s.substring(1))
+         (s(0) + inside, after)
+       }
+       
+       def parse(s: String, nextGroupIndex: Int): (Seq[Node], String, Int)  = if (s.isEmpty) {
+         (Seq(), s, nextGroupIndex)
+       } else {
+         var res = s(0) match {
+         case '(' if s.length >= 3 && s(1) == '?' && (s(2) == '=' || s(2) == '!') =>
+           val (parsed, remaining, newNextGroupIndex) =
+             parse(s.substring(3), nextGroupIndex )
```

Comment:
Taking `substring(x)` repeatedly causes the whole thing to be O(n²). Use an additional `index` parameter to `parse` instead.

Reply:
I removed almost all the calls to substring by adding an explicit parameter pIndex. Renamed s to "pattern", and removed "pattern" from the list of arguments since it is available locally.